### PR TITLE
feat(docker): /container-init.d for advanced initialization

### PR DESCRIPTION
**Pull request by Caian**
This is a proposal to help solve advanced initialization in container environments being discussed in several issues. It follows the model used by postgres container: https://hub.docker.com/_/postgres

This PR adds support for an optional `/container-init.d` directory in the container that can be mounted with custom configuration scripts. These scripts are executed if they have +x permission, otherwise they are just sourced.

This model also allows passing environment variables to the custom scripts when launching the container.

This grants flexibility to and delegates the responsability of bulding complex initializations to the users and administrators.

-----------

License: MIT
Signed-off-by: Caian Benedicto <caianbene@gmail.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -347,6 +347,8 @@ The swarm key initialization can also be done using docker secrets **(requires d
     cat your_swarm.key | docker secret create swarm_key_secret -
     docker run -d --name ipfs_host --secret swarm_key_secret -e IPFS_SWARM_KEY_FILE=/run/secrets/swarm_key_secret -v $ipfs_staging:/export -v $ipfs_data:/data/ipfs -p 4001:4001 -p 127.0.0.1:8080:8080 -p 127.0.0.1:5001:5001 ipfs/go-ipfs:latest
 
+To perform a custom initialization and configuration step, one can create or mount a `/container-init.d` directory containing `.sh` scripts. These scripts are executed if they have `+x` permission, otherwise they are just sourced. The scripts may also depend on environment variables set by the `docker run` command or in `docker-compose.yml`. The custom initialization step is **only** performed during IPFS initialization and is executed **after** the swarm key is copied to the IPFS data directory.
```
Comment (olizilla) [inline README.md]: This looks like a nice pattern! To be more specific here it happens **after** `ipfs init` which might be clearer to folks. It'd be great to show a usage example here too.

Comment (Caian) [reply README.md]: Thank you @olizilla I'm already using this branch to do work so it is not hard to get some examples.

**Comment by Caian** [comment]
README updated with example per request of @olizilla

**Review by lidel — CHANGES_REQUESTED** [review]
I really like how flexible this approach is. :heart:  :+1: 

Agree on the main selling point: hook dir removes the need for things like  https://github.com/ipfs/go-ipfs/pull/8326   and allows people to implement custom initialization without burdening the upstream project  with additional maintenance surface.   

(The mentioned [postgresql/docker-entrypoint.sh](https://github.com/docker-library/postgres/blob/b4b726dbf1885e8e1543526ad9d250fdb2689cbb/14/buster/docker-entrypoint.sh#L147) is very similar in that it executes only during initialization)

### Remaining work

- [x] Rebase on top of latest master (done)
- [ ] Add regression test
  - as this is something people will leverage in their infra, there should be a test in `test/sharness/t0300-docker-image.sh` (or a separate file, if thats easier) that confirms that a script mounted into `/container-init.d` is executed.  It could execute `ipfs bootstrap add` and then we could check if `Bootstrap` section in the config cot  updated.
    - @Caian will you have time to push this over the finish line, or can this be picked up by someone else?

**Comment by BigLep** [comment]
Next step is to add tests.

**Comment by guseggert** [comment]
Yes I believe this is the issue: https://github.com/ipfs/go-ipfs/issues/6101

**Comment by thattommyhall** [comment]
I've closed #8719 as this is more powerful for init. I did consider in that PR longer-lived repos and the fact that you might want to change config at a point other than init, this should allow for that also. 
Adding extra peers is something that you should be able to do w/o re-initialising the repo for instance as a motivating example
I assumed that `ipfs config --json "$name" "$value"` is idempotent so could be used in a startup script, even if you dont add it here I'd like (dis)confirmation of that as I might use the trick in systemd to neaten up our config elsewhere

**Comment by guseggert** [comment]
PR for fixing the Docker sharness tests, which will unblock tests for this: https://github.com/ipfs/go-ipfs/pull/8808

**Comment by guseggert** [comment]
I've written the sharness tests for this, I'm just waiting on https://github.com/ipfs/go-ipfs/pull/8808 to land and then I'll rebase and update this PR

**Comment by thattommyhall** [comment]
Just a quick note to say I've optimistically started using this in our infra, just for local testing at the moment and I love it. The UX is much nicer than my PR as I was forcing a single line of JSON so it was less obvious to :eyes: what's going on.

I think serious prod use will require scripts at startup as well as init though, people need to manage config over time

**Comment by BigLep** [comment]
2022-04-01: 
- Needs review
- Is getting hit by docker flakiness (separate work item).

**Comment by BigLep** [comment]
@thattommyhall : it would be great if you can give review of this as as a user of the functionality.

**Comment by BigLep** [comment]
@thattommyhall : did you meean to close this?

**Review by lidel — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -366,6 +366,48 @@ Basic proof of 'ipfs working' locally:
     # QmT78zSuBmuS4z925WZfrqQ1qHaJ56DQaTfyMUF7F8ff5o
     ipfs cat <that hash>
 
+To perform a custom initialization and configuration step, one can mount `.sh` scripts to a `/container-init.d` directory, or mount the entire directory. These scripts are executed if they have `+x` permission, otherwise they are just sourced. Executed scripts can exit with error and abort the initialization process. The scripts may also depend on environment variables set by the `docker run` command or in `docker-compose.yml`. The custom initialization step is **only** performed during IPFS initialization, **after** `ipfs init`, **after** after the swarm key is copied to the IPFS data directory, and **before** the daemon is started. This also enables custom containers to layer extra initialization steps to the base image by adding more scripts to `/container-init.d`.
```
[the remark this answers is not in the document]
Comment (guseggert) [reply README.md:369]: Yeah it's a good distinction to make, but the prose of this paragraph bothers me a little, let me rewrite it.

[the remark this answers is not in the document]
Comment (guseggert) [reply README.md:369]: I'm going to just remove this and add it to [the IPFS Docker documentation ](http://docs.ipfs.io.ipns.localhost:8080/how-to/run-ipfs-inside-docker/#set-up), which is already linked to from the readme.

**Comment by guseggert** [comment]
I've removed the documentation and will open a PR in https://github.com/ipfs/ipfs-docs w/ docs.

**Review by guseggert — APPROVED** [review]

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lidel) [reply README.md]: Filled https://github.com/ipfs/ipfs-docs/issues/1114

[the remark this answers is not in the document]
Comment (lidel) [reply README.md]: Filled https://github.com/ipfs/ipfs-docs/issues/1114

**Comment by thattommyhall** [comment]
started a repo of interesting examples of this https://github.com/ipfs-shipyard/go-ipfs-docker-examples

**Pull request merged**